File: memaccess.js
const fs = require("fs");
const { WASI } = require("wasi");
const wasi = new WASI();
const importObject = { wasi_snapshot_preview1: wasi.wasiImport };

(async () => {
  const wasm = await WebAssembly.compile(fs.readFileSync("./memaccess.wasm"));
  const instance = await WebAssembly.instantiate(wasm, {
    ...importObject,
    env: {
      transfer_to_runtime: (value, pointer, size) => {
        console.log(
          `(runtime) value=${value} pointer=0x${pointer.toString(
            16
          )} size=${size}`
        );

        const intInMemory = new Uint8Array(
          instance.exports.memory.buffer
        ).slice(pointer, pointer + size);

        console.log(`(memory) value=${new Int32Array(intInMemory.buffer)[0]}`);

        return 0;
      },
    },
  });

  wasi.start(instance);
})();

File: memaccess.c
#include "memaccess.h"
#include <stdio.h>

int main() {
  int my_positive_int = 239234;

  printf("(binary) value=%d pointer=%p size=%lu\n", my_positive_int,
         &my_positive_int, sizeof(my_positive_int));

  transfer_to_runtime(my_positive_int, &my_positive_int,
                      sizeof(my_positive_int));

  int my_negative_int = -239234;

  printf("(binary) value=%d pointer=%p size=%lu\n", my_negative_int,
         &my_negative_int, sizeof(my_negative_int));

  transfer_to_runtime(my_negative_int, &my_negative_int,
                      sizeof(my_negative_int));

  return 0;
}